soh/soh/Enhancements/randomizer/location_access.h

```diff
@@ -232,92 +233,80 @@ class Region {
-     * - Going back into big block room is complex, but the only check there is child only so not a concern
-     * - Everything else is possible with basic adult movement, or is impossible for child to reach glitchlessly
-     * Anything 7 key locked does not need to be checked as shared, as all child access is Certain and because of this
-     workaround we don't need to fake Adult access, meaning that is also Certain.
-     * All of this combined means that when checking if adult can reach a location in universe 2, we only have to ask if
-     it is a 6 key locked location or not.
-
-     * Knowing all of this this, we can confirm things are logical in 3 different ways:
-     * - If we have Adult Access, we know it is Certain Access, so they can get checks alone.
-     * - If we have 7 keys, child has Certain Access as we know they cannot be locked out, so can get checks alone,
-     otherwise we check the logical overlap
-     * - If Child and Adult can get the check (ignoring actual adult access to the location), and the location is either
-     not 6 key locked or we have 6 keys, we can get the check with the overlap
+     * This logic covers checks that exist in the shared areas of Spirit from a glitchless standpoint.
+     * This code will fail if any glitch allows Adult to go in the Child spirit door first or vice versa as it relies on
+     specific ages
+
+     * There are 4 possibilities for passing a check, but first I have to talk about parallel universes.
+
+     * In the first universe, the player enters spirit as child, and spends as many keys as they can to lock adult out
+     * In the second, they enter as adult and spend as many keys as they can to lock child out.
+
+     * When an Age can no longer be kept out by the other age, that age is said to have Certain Access to a region
+     * If both ages have access to a region with a certain number of keys, but there is no Certain Access,
+     * then a check is only in logic if both ages can collect the check independently
+
+     * If an age has Certain Access then that age can collect checks alone,
+     * and there is no reason to check the other age untile the universes converge.
+
+     * The universes converge when the player has all the keys, giving both ages Certain Access.
+
+     * We must check for these universes manually as we allow technical access with minimum keys for
+     * technical reasons as otherwise the code will never run
      */
-    bool MQSpiritShared(ConditionFn condition, bool IsBrokenWall, bool anyAge = false) {
-        // if we have Certain Access as child, we can check anyAge and if true, resolve a condition with Here as if
-        // adult is here it's also Certain Access
-        if (logic->SmallKeys(RR_SPIRIT_TEMPLE, 7)) {
+
+    bool SpiritShared(ConditionFn condition, ConditionFn childAccess, ConditionFn adultAccess, uint8_t adultKeys,
+                      uint8_t childKeys, uint8_t eitherKeys, bool anyAge = false) {
```

Comment:
confusing having swapped order of child fn / adult fn, but adult keys / child keys